src/layouts/app/translations-mixin.ts

```diff
@@ -16,60 +21,86 @@ export default (superClass: Constructor<LitElement & HassBaseEl>) =>
     protected firstUpdated(changedProps) {
       super.firstUpdated(changedProps);
       this.addEventListener("hass-language-select", (e) =>
-        this._selectLanguage(e)
+        this._selectLanguage((e as CustomEvent).detail.language, true)
       );
-      this._loadResources();
+      this._loadCoreTranslations(getLocalLanguage());
     }
 
     protected hassConnected() {
       super.hassConnected();
-      this._loadBackendTranslations();
-      this.style.direction = computeRTL(this.hass!) ? "rtl" : "ltr";
+      getUserLanguage(this.hass!).then((language) => {
+        if (language && this.hass!.language !== language) {
+          // We just get language from backend, no need to save back
+          this._selectLanguage(language, false);
+        }
+      });
+      this._loadTranslations(this.hass!);
     }
 
     protected hassReconnected() {
       super.hassReconnected();
-      this._loadBackendTranslations();
+      this._loadTranslations(this.hass!);
     }
 
     protected panelUrlChanged(newPanelUrl) {
       super.panelUrlChanged(newPanelUrl);
-      this._loadTranslationFragment(newPanelUrl);
+      // this may be triggered before hassConnected
+      this._loadFragmentTranslations(
+        this.hass ? this.hass.language : getLocalLanguage(),
+        newPanelUrl
+      );
     }
 
-    private async _loadBackendTranslations() {
-      const hass = this.hass;
-      if (!hass || !hass.language) {
+    private _selectLanguage(language: string, saveToBackend: boolean) {
+      if (!this.hass) {
+        // should not happen, do it to avoid use this.hass!
         return;
       }
 
-      const language = hass.selectedLanguage || hass.language;
+      // update selectedLanguage so that it can be saved to local storage
+      this._updateHass({ language, selectedLanguage: language });
+      storeState(this.hass);
+      if (saveToBackend) {
+        saveFrontendUserData(this.hass, "language", { language });
+      }
 
-      const { resources } = await hass.callWS({
-        type: "frontend/get_translations",
-        language,
-      });
+      this._loadTranslations(this.hass);
+    }
+
+    private _loadTranslations(hass: HomeAssistant) {
+      this.style.direction = computeRTL(hass) ? "rtl" : "ltr";
```

Comment:
It feels weird to do it in a method that is called `loadTranslations`. Let's rename it to `_applyTranslations` ?